File: backend/pkg/gameplugin/sdk/game-plugin/src/control.ts
import { Client, MessageOption } from './client';

export const METHOD_AUTHORITY_SNAPSHOT = 'control.authority.snapshot';
export const METHOD_CONTROL_REGISTER_SINK = 'control.sink.register';
export const METHOD_CONTROL_OUTPUT = 'control.output';
export const METHOD_CONTROL_AUTHORITY_TAKEOVER = 'control.authority.takeover';
export const METHOD_CONTROL_AUTHORITY_RELEASE = 'control.authority.release';
export const METHOD_EMERGENCY_STOP = 'emergency.stop';
export const METHOD_CONTROL_SINK_DISPATCH = 'control.sink.dispatch';

export const CONTROL_MODE_OBSERVE_ONLY = 'observe_only';
export const CONTROL_MODE_ASSIST = 'assist';
export const CONTROL_MODE_SHARED_CONTROL = 'shared_control';
export const CONTROL_MODE_PLUGIN_CONTROL = 'plugin_control';
export const CONTROL_MODE_USER_CONTROL = 'user_control';
export const CONTROL_MODE_SUSPENDED = 'suspended';

// Backward-compatible symbol names using canonical GameHost v1 wire values.
export const CONTROL_MODE_OBSERVE = CONTROL_MODE_OBSERVE_ONLY;
export const CONTROL_MODE_SHARED = CONTROL_MODE_SHARED_CONTROL;
export const CONTROL_MODE_PLUGIN = CONTROL_MODE_PLUGIN_CONTROL;
export const CONTROL_MODE_USER = CONTROL_MODE_USER_CONTROL;

export const OUTPUT_KIND_CUSTOM_RPC = 'custom_rpc';
export const OUTPUT_KIND_CHANNEL = 'channel';
export const OUTPUT_KIND_BINARY = 'binary';
export const OUTPUT_KIND_EFFECT = 'effect';

export interface AuthoritySnapshot {
  runtimeId: string;
  pluginId: string;
  mode: string;
  epoch: number;
  updatedAt: string;
  lastTransitionReason?: string;
  lastTransitionActor?: string;
}

export interface ControlSinkRegisterInput {
  sinkId: string;
  kind: string;
  metadata?: Record<string, unknown>;
}

export interface ControlSinkRegisterResult {
  sinkId: string;
  registered: boolean;
}

export interface ControlOutputInput {
  outputId: string;
  sinkId: string;
  epoch: number;
  payload: unknown;
}

export interface ControlOutputResult {
  outputId: string;
  allowed: boolean;
  reason?: string;
  currentEpoch: number;
  generation: number;
}

export interface AuthorityTakeoverInput {
  expectedEpoch?: number;
}

export interface AuthorityTakeoverResult {
  previousMode: string;
  newMode: string;
  previousEpoch: number;
  newEpoch: number;
  snapshot: AuthoritySnapshot;
}

export interface AuthorityReleaseInput {
  targetMode?: string;
  expectedEpoch?: number;
}

export interface AuthorityReleaseResult {
  previousMode: string;
  newMode: string;
  previousEpoch: number;
  newEpoch: number;
  snapshot: AuthoritySnapshot;
}

export interface EmergencyStopInput {
  idempotencyKey?: string;
}

export interface EmergencyStopResult {
  operationId: string;
  runtimeId: string;
  state: string;
  actor: string;
  reason: string;
  success: boolean;
  criticalFailure: boolean;
  residue?: string[];
  startedAt: string;
  finishedAt: string;
}

export interface SinkEffectDispatchPayload {
  sinkId: string;
  serviceId: string;
  outputId: string;
  epoch: number;
  generation: number;
  payload: unknown;
}

export async function getAuthoritySnapshot(
  client: Client,
  opts: MessageOption[] = []
): Promise<AuthoritySnapshot> {
  const envelope = await client.sendReservedRequest(METHOD_AUTHORITY_SNAPSHOT, {}, ...opts);
  return envelope.payload as AuthoritySnapshot;
}

export async function registerControlSink(
  client: Client,
  input: ControlSinkRegisterInput,
  opts: MessageOption[] = []
): Promise<ControlSinkRegisterResult> {
  const envelope = await client.sendReservedRequest(METHOD_CONTROL_REGISTER_SINK, input, ...opts);
  return envelope.payload as ControlSinkRegisterResult;
}

export async function submitControlOutput(
  client: Client,
  input: ControlOutputInput,
  opts: MessageOption[] = []
): Promise<ControlOutputResult> {
  const envelope = await client.sendReservedRequest(METHOD_CONTROL_OUTPUT, input, ...opts);
  return envelope.payload as ControlOutputResult;
}

export async function takeoverAuthority(
  client: Client,
  input: AuthorityTakeoverInput = {},
  opts: MessageOption[] = []
): Promise<AuthorityTakeoverResult> {
  const envelope = await client.sendReservedRequest(METHOD_CONTROL_AUTHORITY_TAKEOVER, input, ...opts);
  return envelope.payload as AuthorityTakeoverResult;
}

export async function releaseAuthority(
  client: Client,
  input: AuthorityReleaseInput = {},
  opts: MessageOption[] = []
): Promise<AuthorityReleaseResult> {
  const envelope = await client.sendReservedRequest(METHOD_CONTROL_AUTHORITY_RELEASE, input, ...opts);
  return envelope.payload as AuthorityReleaseResult;
}

export async function emergencyStop(
  client: Client,
  input: EmergencyStopInput = {},
  opts: MessageOption[] = []
): Promise<EmergencyStopResult> {
  const envelope = await client.sendReservedRequest(METHOD_EMERGENCY_STOP, input, ...opts);
  return envelope.payload as EmergencyStopResult;
}

export interface SinkEffectCommitResult {
  accepted: boolean;
  committed: boolean;
  effectId: string;
  generation: number;
  errorCode?: string;
  message?: string;
}

export type SinkDispatchHandler = (payload: SinkEffectDispatchPayload) => Promise<SinkEffectCommitResult>;

export function registerSinkDispatchHandler(
  registry: import('./handler').HandlerRegistry,
  handler: SinkDispatchHandler
): void {
  registry.registerRequest(METHOD_CONTROL_SINK_DISPATCH, async (request) => {
    const payload = request.payload as SinkEffectDispatchPayload;
    const result = await handler(payload);
    return result;
  });
}

File: backend/pkg/gameplugin/sdk/go/control.go
package sdk

import (
	"context"
	"encoding/json"
	"sync"
	"time"

	"github.com/u-ai/backend/pkg/gameplugin/protocol/contracts"
)

const (
	MethodAuthoritySnapshot   = "control.authority.snapshot"
	MethodControlRegisterSink = "control.sink.register"
	MethodControlOutput       = "control.output"
	MethodControlAuthorityTakeover = "control.authority.takeover"
	MethodControlAuthorityRelease  = "control.authority.release"
	MethodEmergencyStopInitiate   = "emergency.stop.initiate"
	MethodEmergencyStopStatus     = "emergency.stop.status"
)

const (
	ControlModeObserve     = "observe"
	ControlModeAssist      = "assist"
	ControlModeShared      = "shared"
	ControlModePlugin      = "plugin"
	ControlModeUser        = "user"
	ControlModeSuspended   = "suspended"
)

const (
	OutputKindCustomRPC = "custom_rpc"
	OutputKindChannel   = "channel"
	OutputKindBinary    = "binary"
	OutputKindEffect    = "effect"
)

const (
	OutputDeniedStaleEpoch      = "stale_epoch"
	OutputDeniedPermission      = "permission_denied"
	OutputDeniedAuthorityMode   = "authority_mode_denied"
	OutputDeniedHostPolicy      = "host_policy_denied"
	OutputDeniedGateClosed      = "gate_closed"
	OutputDeniedRuntimeNotFound = "runtime_not_found"
	OutputDeniedServiceNotFound = "service_not_found"
	OutputDeniedInvalidPeer     = "invalid_peer"
	OutputDeniedNotEligible     = "runtime_not_eligible"
	OutputDeniedNotReady        = "not_ready"
)

const (
	TakeoverActorUser   = "user"
	TakeoverActorHost   = "host"
	TakeoverActorSystem = "system"
)

const (
	EmergencyStopStateRequested         = "requested"
	EmergencyStopStateClosingGate       = "closing_gate"
	EmergencyStopStateSuspending        = "suspending"
	EmergencyStopStateCancellingWork    = "cancelling_work"
	EmergencyStopStateStoppingRuntime   = "stopping_runtime"
	EmergencyStopStateClosingConnection = "closing_connection"
	EmergencyStopStateRevokingLeases    = "revoking_leases"
	EmergencyStopStateCleaningResources = "cleaning_resources"
	EmergencyStopStateVerifying         = "verifying"
	EmergencyStopStateCompleted         = "completed"
	EmergencyStopStateFailed            = "failed"
)

type AuthoritySnapshot struct {
	RuntimeID string `json:"runtimeId"`
	PluginID  string `json:"pluginId"`
	Mode      string `json:"mode"`
	Epoch     uint64 `json:"epoch"`
	ServiceID string `json:"serviceId,omitempty"`
	UpdatedAt int64  `json:"updatedAt,omitempty"`
	Valid     bool   `json:"valid"`
}

type ControlSinkRegisterInput = contracts.SinkRegisterInput

type ControlSinkRegisterResult struct {
	SinkID  string `json:"sinkId"`
	Registered bool `json:"registered"`
}

type ControlOutputInput = contracts.ControlOutputInput

type ControlOutputResult = contracts.ControlOutputResult

type AuthorityTakeoverInput struct {
	TargetMode    string `json:"targetMode"`
	Actor         string `json:"actor"`
	ExpectedEpoch uint64 `json:"expectedEpoch,omitempty"`
	ServiceID     string `json:"serviceId,omitempty"`
}

type AuthorityTakeoverResult struct {
	PreviousMode  string `json:"previousMode"`
	NewMode       string `json:"newMode"`
	PreviousEpoch uint64 `json:"previousEpoch"`
	NewEpoch      uint64 `json:"newEpoch"`
	Success       bool   `json:"success"`
	Reason        string `json:"reason,omitempty"`
}

type AuthorityReleaseInput struct {
	TargetMode    string `json:"targetMode"`
	Actor         string `json:"actor"`
	ExpectedEpoch uint64 `json:"expectedEpoch,omitempty"`
	ServiceID     string `json:"serviceId,omitempty"`
}

type AuthorityReleaseResult struct {
	PreviousMode  string `json:"previousMode"`
	NewMode       string `json:"newMode"`
	PreviousEpoch uint64 `json:"previousEpoch"`
	NewEpoch      uint64 `json:"newEpoch"`
	Success       bool   `json:"success"`
	Reason        string `json:"reason,omitempty"`
}

type EmergencyStopStatusInput struct {
	OperationID string `json:"operationId,omitempty"`
}

type EmergencyStopStatusResult struct {
	OperationID   string `json:"operationId"`
	State         string `json:"state"`
	Active        bool   `json:"active"`
	Reason        string `json:"reason,omitempty"`
	InitiatedAt   int64  `json:"initiatedAt,omitempty"`
	CompletedAt   int64  `json:"completedAt,omitempty"`
}

func (c *Client) GetAuthoritySnapshot(ctx context.Context, runtimeID, serviceID string, opts ...MessageOption) (AuthoritySnapshot, error) {
	input := map[string]any{
		"runtimeId": runtimeID,
		"serviceId": serviceID,
	}
	envelope, err := c.SendReservedRequest(ctx, MethodAuthoritySnapshot, input, opts...)
	if err != nil {
		return AuthoritySnapshot{}, err
	}
	var out AuthoritySnapshot
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return AuthoritySnapshot{}, NewEncodeError("unmarshal authority snapshot: %v", err)
		}
	}
	return out, nil
}

func (c *Client) RegisterControlSink(ctx context.Context, input ControlSinkRegisterInput, opts ...MessageOption) (ControlSinkRegisterResult, error) {
	envelope, err := c.SendReservedRequest(ctx, MethodControlRegisterSink, input, opts...)
	if err != nil {
		return ControlSinkRegisterResult{}, err
	}
	var out ControlSinkRegisterResult
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return ControlSinkRegisterResult{}, NewEncodeError("unmarshal control sink register response: %v", err)
		}
	}
	return out, nil
}

func (c *Client) SubmitControlOutput(ctx context.Context, input ControlOutputInput, opts ...MessageOption) (ControlOutputResult, error) {
	envelope, err := c.SendReservedRequest(ctx, MethodControlOutput, input, opts...)
	if err != nil {
		return ControlOutputResult{}, err
	}
	var out ControlOutputResult
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return ControlOutputResult{}, NewEncodeError("unmarshal control output response: %v", err)
		}
	}
	return out, nil
}

func (c *Client) TakeoverAuthority(ctx context.Context, input AuthorityTakeoverInput, runtimeID string, opts ...MessageOption) (AuthorityTakeoverResult, error) {
	envelope, err := c.SendReservedRequest(ctx, MethodControlAuthorityTakeover, input, opts...)
	if err != nil {
		return AuthorityTakeoverResult{}, err
	}
	var out AuthorityTakeoverResult
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return AuthorityTakeoverResult{}, NewEncodeError("unmarshal authority takeover response: %v", err)
		}
	}
	return out, nil
}

func (c *Client) ReleaseAuthority(ctx context.Context, input AuthorityReleaseInput, runtimeID string, opts ...MessageOption) (AuthorityReleaseResult, error) {
	envelope, err := c.SendReservedRequest(ctx, MethodControlAuthorityRelease, input, opts...)
	if err != nil {
		return AuthorityReleaseResult{}, err
	}
	var out AuthorityReleaseResult
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return AuthorityReleaseResult{}, NewEncodeError("unmarshal authority release response: %v", err)
		}
	}
	return out, nil
}

func (c *Client) GetEmergencyStopStatus(ctx context.Context, input EmergencyStopStatusInput, opts ...MessageOption) (EmergencyStopStatusResult, error) {
	envelope, err := c.SendReservedRequest(ctx, MethodEmergencyStopStatus, input, opts...)
	if err != nil {
		return EmergencyStopStatusResult{}, err
	}
	var out EmergencyStopStatusResult
	if len(envelope.Payload) > 0 {
		if err := json.Unmarshal(envelope.Payload, &out); err != nil {
			return EmergencyStopStatusResult{}, NewEncodeError("unmarshal emergency stop status: %v", err)
		}
	}
	return out, nil
}

const EffectCommitCacheTTL = 5 * time.Minute

type EffectCommitCacheEntry struct {
	Result    SinkEffectCommitResult
	Timestamp time.Time
}

type EffectCommitCache struct {
	mu      sync.Mutex
	entries map[string]EffectCommitCacheEntry
	ttl     time.Duration
}

func NewEffectCommitCache() *EffectCommitCache {
	return &EffectCommitCache{
		entries: make(map[string]EffectCommitCacheEntry),
		ttl:     EffectCommitCacheTTL,
	}
}

func (c *EffectCommitCache) Get(outputID string) (SinkEffectCommitResult, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.entries[outputID]
	if !ok {
		return SinkEffectCommitResult{}, false
	}
	if time.Since(entry.Timestamp) > c.ttl {
		delete(c.entries, outputID)
		return SinkEffectCommitResult{}, false
	}
	return entry.Result, true
}

func (c *EffectCommitCache) Put(outputID string, result SinkEffectCommitResult) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries[outputID] = EffectCommitCacheEntry{
		Result:    result,
		Timestamp: time.Now(),
	}
}

func (c *EffectCommitCache) Clear() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.entries = make(map[string]EffectCommitCacheEntry)
}

type SinkEffectDispatchPayload = contracts.SinkEffectDispatchPayload

type SinkEffectCommitResult = contracts.SinkEffectCommitResult

type SinkDispatchHandler func(payload SinkEffectDispatchPayload) (SinkEffectCommitResult, error)

func RegisterSinkDispatchHandler(registry interface {
	Register(method string, handler func(request interface{}) (interface{}, error))
}, handler SinkDispatchHandler, cache *EffectCommitCache) {
	registry.Register(MethodSinkDispatch, func(request interface{}) (interface{}, error) {
		payload, ok := request.(SinkEffectDispatchPayload)
		if !ok {
			return SinkEffectCommitResult{
				Accepted:   false,
				Committed:  false,
				EffectID:   "",
				Generation: 0,
				ErrorCode:  "invalid_argument",
				Message:    "invalid sink dispatch payload",
			}, nil
		}
		if cache != nil {
			if cached, hit := cache.Get(payload.OutputID); hit {
				return cached, nil
			}
		}
		result, err := handler(payload)
		if err != nil {
			return SinkEffectCommitResult{
				Accepted:   false,
				Committed:  false,
				EffectID:   "",
				Generation: 0,
				ErrorCode:  "handler_error",
				Message:    err.Error(),
			}, nil
		}
		if cache != nil && result.Committed {
			cache.Put(payload.OutputID, result)
		}
		return result, nil
	})
}

const MethodSinkDispatch = contracts.MethodSinkDispatch
